Sync wait in RemoteLayerWithRemoteRenderingBackingStoreCollection::prepareBackingStoresForDisplay delays rendering.

mattwoodrow opened the pull request:
#### 3dd074a9e5d206e613ec48c3d3f57388f00c62c6
<pre>
Sync wait in RemoteLayerWithRemoteRenderingBackingStoreCollection::prepareBackingStoresForDisplay delays rendering.
<a href="https://bugs.webkit.org/show_bug.cgi?id=265295">https://bugs.webkit.org/show_bug.cgi?id=265295</a>
&lt;<a href="https://rdar.apple.com/116792212">rdar://116792212</a>&gt;

Reviewed by Kimmo Kinnunen.

In order to start issuing rendering commands for a layer, we need to determine which ImageBuffer is not
in-use and can be drawn to. We also need to know if the existing pixel contents have been discarded due
to being marked volatile. Both of these checks can only be done in the GPU process, so the previous
behaviour was to have a synchronous call to prepareBackingStoresForDisplay at the start of rendering.

This changes makes that process asynchronous (in the common case), by moving management of the buffer
triplet (front, back, secondary back) to the GPU process, and making the choice of which buffer is used
opaque from the perspective of the web process.

For each frame, the WebProcess allocates a new image buffer identifier (via RemoteDisplayListRecorderProxy),
and the GPU process asynchronously ensures that this is mapped to the right ImageBuffer (either by picking
an available one, or allocating if needed, similar to how recycling IOSurfaces from the pool happens). It
also queues copying of the pixels from the previous frame, clipping to the new draw region, and clearing the
draw region.

The only case where this isn&apos;t possible is when the existing front buffer has been marked volatile (and
might be discarded). This should be relatively rare (and not happen during high performance animations),
and we add a sync wait to determine if the full frame needs to be redrawn or not.

This adds a new RemoteRenderingBufferSet object in the GPU process that manages the set of buffers, and
preparing backing stores, volatility setting and flushing are changed to happen in terms of the buffer set
rather than individual ImageBuffers.

We unfortuntely still have to have a main-thread sync wait at the end of the rendering update, in order to
get the mach_port_t/IOSurface handle for the front buffer used. We also need to get the set of
RenderingResourceIdentifiers used, since these are the keys for the CAIOSurface caching in the UI process.
In the common case, this message will be sent back while the WebProcess is issuing drawing commands, and
the sync wait will process the pending message immediately.
Ideally a future change will handle sending IOSurfaces (via mach_port_t) from GPUP to UI process directly,
which would remove the need for this.

The current state is a performance improvement, since we can issue rendering commands before any sync
waits, but an ideal future would remove the sync waiting from rendering entirely.

This also fixes a bug with volatility, where we can send a volatility request, then do a sync prepare for
display (which sets non volatile), and finally process the asynchronous reply from the volatility request.
In that case the web process would cache locally that the buffer is volatile and not make further requests
for it. This is fixed by tracking transaction identifiers for volatility requests to ensure that we don&apos;t
get ordering problems.

* Source/WTF/wtf/text/TextStream.h:
(WTF::operator&lt;&lt;):
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.cpp:
(WebKit::RemoteImageBuffer::create):
(WebKit::RemoteImageBuffer::RemoteImageBuffer):
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.h:
(WebKit::RemoteImageBuffer::identifier const):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::didFailCreateImageBuffer):
(WebKit::RemoteRenderingBackend::didCreateImageBuffer):
(WebKit::RemoteRenderingBackend::moveToImageBuffer):
(WebKit::RemoteRenderingBackend::createImageBuffer):
(WebKit::RemoteRenderingBackend::createRemoteRenderingBufferSet):
(WebKit::RemoteRenderingBackend::releaseRemoteRenderingBufferSet):
(WebKit::RemoteRenderingBackend::updateRemoteRenderingBufferSetConfiguration):
(WebKit::RemoteRenderingBackend::createBuffersForDisplay):
(WebKit::RemoteRenderingBackend::markSurfacesVolatile):
(WebKit::handleFromBuffer): Deleted.
(WebKit::RemoteRenderingBackend::prepareBuffersForDisplay): Deleted.
(WebKit::RemoteRenderingBackend::prepareLayerBuffersForDisplay): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBufferSet.cpp: Added.
(WebKit::RemoteRenderingBufferSet::updateConfiguration):
(WebKit::RemoteRenderingBufferSet::ensureFrontBufferForDisplay):
(WebKit::RemoteRenderingBufferSet::prepareFrontBufferForDisplay):
(WebKit::RemoteRenderingBufferSet::makeBuffersVolatile):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBufferSet.h: Added.
(WebKit::RemoteRenderingBufferSet::isOpaque const):
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
(IPC::StreamClientConnection::waitForAsyncReplyAndDispatchImmediately):
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
(class_template_headers):
(headers_for_type):
* Source/WebKit/Shared/RemoteBufferSetIdentifier.h: Copied from Source/WebKit/WebProcess/GPU/graphics/BufferIdentifierSet.h.
* Source/WebKit/Shared/RemoteLayerTree/BufferAndBackendInfo.h: Removed.
* Source/WebKit/Shared/RemoteLayerTree/BufferAndBackendInfo.serialization.in: Removed.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
(WebKit::RemoteLayerBackingStore::dirtyRegion):
(WebKit::RemoteLayerBackingStore::setBufferCacheIdentifiers):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::ensureBackingStore):
(WebKit::RemoteLayerBackingStore::encode const):
(WebKit::RemoteLayerBackingStoreProperties::decode):
(WebKit::RemoteLayerBackingStoreProperties::dump const):
(WebKit::RemoteLayerBackingStore::setFrontBuffer):
(WebKit::RemoteLayerBackingStore::supportsPartialRepaint const):
(WebKit::RemoteLayerBackingStore::drawingRequiresClearedPixels const):
(WebKit::RemoteLayerBackingStore::setDelegatedContents):
(WebKit::RemoteLayerBackingStore::paintContents):
(WebKit::RemoteLayerBackingStore::drawInContext):
(WebKit::RemoteLayerBackingStoreProperties::applyBackingStoreToLayer):
(WebKit::RemoteLayerBackingStoreProperties::updateCachedBuffers):
(WebKit::RemoteLayerBackingStore::applySwappedBuffers): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.h:
(WebKit::RemoteLayerBackingStoreCollection::backingStoreUpdated):
(WebKit::RemoteLayerBackingStoreCollection::backingStoreWillBeEncoded):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStoreCollection.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStoreCollection.mm:
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::paintReachableBackingStoreContents):
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::prepareBackingStoresForDisplay):
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::allocateBufferForBackingStore):
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::collectBackingStoreBufferIdentifiersToMarkVolatile):
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::collectAllBufferIdentifiersToMarkVolatile):
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::markBackingStoreVolatileAfterReachabilityChange):
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::tryMarkAllBackingStoreVolatile):
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::markAllBackingStoreVolatileFromTimer):
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::sendMarkBuffersVolatile):
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::gpuProcessConnectionWasDestroyed):
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::backingStoreWasCreated):
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::backingStoreWillBeDestroyed):
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::backingStoreUpdated):
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::backingStoreWillBeEncoded):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.messages.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::asyncSetLayerContents):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::asyncSetLayerContents):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.h:
(WebKit::RemoteLayerTreeNode::asyncContentsIdentifier const):
(WebKit::RemoteLayerTreeNode::setAsyncContentsIdentifier):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/BufferIdentifierSet.h:
(WebKit::operator&lt;&lt;):
* Source/WebKit/WebProcess/GPU/graphics/BufferIdentifierSet.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/PrepareBackingStoreBuffersData.cpp:
(WebKit::operator&lt;&lt;):
* Source/WebKit/WebProcess/GPU/graphics/PrepareBackingStoreBuffersData.h:
* Source/WebKit/WebProcess/GPU/graphics/PrepareBackingStoreBuffersData.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::RemoteImageBufferProxy):
(WebKit::m_remoteBufferManager):
(WebKit::RemoteImageBufferProxy::remoteBufferWasDestroyed):
(WebKit::RemoteSerializedImageBufferProxy::sinkIntoImageBuffer):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
(WebKit::RemoteImageBufferProxy::create):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::didClose):
(WebKit::RemoteRenderingBackendProxy::createImageBuffer):
(WebKit::RemoteRenderingBackendProxy::createRemoteRenderingBufferSet):
(WebKit::RemoteRenderingBackendProxy::releaseRemoteRenderingBufferSet):
(WebKit::RemoteRenderingBackendProxy::updateRemoteRenderingBufferSetConfiguration):
(WebKit::RemoteRenderingBackendProxy::prepareBuffersForDisplay):
(WebKit::RemoteRenderingBackendProxy::ensurePrepareCompleted):
(WebKit::RemoteRenderingBackendProxy::markSurfacesVolatile):
(WebKit::RemoteRenderingBackendProxy::didMarkLayersAsVolatile):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBufferSetProxy.cpp: Copied from Source/WebKit/WebProcess/GPU/graphics/BufferIdentifierSet.h.
(WebKit::RemoteRenderingBufferSetProxy::RemoteRenderingBufferSetProxy):
(WebKit::RemoteRenderingBufferSetProxy::~RemoteRenderingBufferSetProxy):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBufferSetProxy.h: Added.
(WebKit::RemoteRenderingBufferSetProxy::remoteBufferSetIdentifier const):
(WebKit::RemoteRenderingBufferSetProxy::requestedVolatility):
(WebKit::RemoteRenderingBufferSetProxy::confirmedVolatility):
(WebKit::RemoteRenderingBufferSetProxy::addRequestedVolatility):
(WebKit::RemoteRenderingBufferSetProxy::setConfirmedVolatility):
(WebKit::RemoteRenderingBufferSetProxy::clearVolatilityUntilAfter):
(WebKit::operator&lt;&lt;):
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::remoteResourceCacheWasDestroyed):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDocumentLayout.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.mm:

Canonical link: <a href="https://commits.webkit.org/271747@main">https://commits.webkit.org/271747@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8de9bb0f71c0eaf1f19e02c98abf1377e09e58ea

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29522 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8189 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30846 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32060 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/26753 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/10338 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5482 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29795 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6840 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/25228 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5856 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/29676 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/5979 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/26314 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33404 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/25389 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26944 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26734 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/32210 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/29761 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/5937 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/4127 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/29984 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7669 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/36094 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7008 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/6476 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7779 "Passed tests") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/6460 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
```diff
@@ -255,6 +259,28 @@ void RemoteRenderingBackend::releaseImageBuffer(RenderingResourceIdentifier rend
     MESSAGE_CHECK(success, "Resource is being released before being cached."_s);
 }
 
+void RemoteRenderingBackend::createRemoteRenderingBufferSet(RemoteBufferSetIdentifier bufferSetIdentifier)
```
RemoteRenderingBackend message CreateImageBuffer creates a remote ImageBuffer

RemoteRenderingBackend message createImageBufferSet could create a remote ImageBufferSet

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
```diff
@@ -255,6 +259,28 @@ void RemoteRenderingBackend::releaseImageBuffer(RenderingResourceIdentifier rend
+    MESSAGE_CHECK(success, "BufferSet is being released before being created"_s);
+}
+
+void RemoteRenderingBackend::updateRemoteRenderingBufferSetConfiguration(RemoteBufferSetIdentifier bufferSetIdentifier, const FloatSize& logicalSize, RenderingMode renderingMode, float resolutionScale, const DestinationColorSpace& colorSpace, PixelFormat pixelFormat)
```
This is a updateConfiguration message to a remote ImageBufferSet identified by identifier
It could be defined in RemoteBufferSet.messages.in.

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/PrepareBackingStoreBuffersData.cpp:
```diff
@@ -34,21 +34,15 @@
 TextStream& operator<<(TextStream& ts, const PrepareBackingStoreBuffersInputData& inputData)
 {
-    ts << inputData.bufferSet << ", supportsPartialRepaint: " << inputData.supportsPartialRepaint << " hasEmptyDirtyRegion: " << inputData.hasEmptyDirtyRegion;
+    ts << "supportsPartialRepaint: " << inputData.supportsPartialRepaint << " hasEmptyDirtyRegion: " << inputData.hasEmptyDirtyRegion;
```
Did we lose the ability to dump the inputData.remoteBufferSet, newFrontBuffer?

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBufferSetProxy.h:
```diff
@@ -0,0 +1,86 @@
+    OptionSet<BufferInSetType> confirmedVolatility() { return m_confirmedVolatility; }
+
+private:
+    void addRequestedVolatility(OptionSet<BufferInSetType> request)
```
These could be just normal functions in .cpp file.

kkinnunen-apple commented on Source/WTF/wtf/text/TextStream.h:
```diff
@@ -349,6 +349,18 @@ TextStream& operator<<(TextStream& ts, const std::array<T, size>& array)
+template<typename T, typename U>
+TextStream& operator<<(TextStream& ts, const std::pair<T, U>& pair)
+{
+    ts << "[";
```
Opinion  nit: I don't think this function is as complex as to warrant grouping whitespace to clarify different sections of the code?
I'd find following the easiest to read:
`return ts << "[" << pair.first << ", " pair.second << "]";`

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.h:
```diff
@@ -73,6 +73,7 @@ class RemoteImageBuffer : public IPC::StreamMessageReceiver {
     Ref<WebCore::ImageBuffer> m_imageBuffer;
     ScopedRenderingResourcesRequest m_renderingResourcesRequest { ScopedRenderingResourcesRequest::acquire() };
     std::optional<IPC::Signal> m_flushSignal;
+    WebCore::RenderingResourceIdentifier m_renderingResourceIdentifier;
```
Was there a reason the image buffer identifier does not work, as it does today?

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/BufferIdentifierSet.h:
```diff
@@ -27,16 +27,53 @@
+    SecondaryBack = 1 << 2,
+};
+
+inline TextStream& operator<<(TextStream& ts, BufferInSetType& bufferType)
```
no ref&

kkinnunen-apple commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -324,7 +324,7 @@
 }
 
-void RemoteLayerTreeDrawingAreaProxy::asyncSetLayerContents(WebCore::PlatformLayerIdentifier layerID, ImageBufferBackendHandle&& handle, const WebCore::RenderingResourceIdentifier& identifier)
+void RemoteLayerTreeDrawingAreaProxy::asyncSetLayerContents(WebCore::PlatformLayerIdentifier layerID, ImageBufferBackendHandle&& handle, const WebCore::ProcessQualified<WebCore::RenderingResourceIdentifier>& identifier)
```
I'm not really fan of transferring ProcessQualified.
I don't immediately understand why these changes are being,  I don't see the use site.
If it is needed, the UI process should resolve it based on the connection the message came, not like this.

mattwoodrow replied:
This one is a bit sad, admittedly, though hopefully can be fixed.

The UI process is keeping cached `CAIOSurface` objects for the three buffers, and needs to know which of these to keep/recycle/allocate each frame. It does so by tracking the `RenderingResourceIdentifier` associated with each, and the layer transaction provides the new three identifiers (plus one actual buffer handle) for it to match/update the cache.

Since the WebProcess no longer manages the three buffers, it doesn't have those identifiers to send. I made the GPUP send back the `BufferIdentifierSet` as part of the async reply just for this purpose. That set uses `ProcessQualified` (with the GPUP's process identity), since the identifiers were allocated in the GPUP, and we don't want them to collide with values allocated in the WebProcess (or in a different GPUP if the previous one crashes).

This code is using identifiers to resolve races between buffers delivered from `OffscreenCanvas` in a worker thread, and the same buffers being delivered through the main thread transaction, and doesn't really need the above. Unfortunately the main-thread path is using the same serialisation as the process qualified cache identifiers, so this changes to match.

It's a bit of sadness from RemoteLayerBackingStore representing quite different things (in-process 3 buffers, GPUP 3 buffers, or a delegated buffer) but all with a single serialization.

Maybe I'll just duplicate the identifier serialisation a bit rather than making this change.

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteRenderingBufferSet.cpp:
```diff
@@ -0,0 +1,200 @@
+            needsFullDisplay = true;
+    }
+
+    if (!m_frontBuffer || !inputData.supportsPartialRepaint || isSmallLayerBacking(m_frontBuffer->parameters()))
```
This isSmalllayerBacking function got lost, unified build saves you today but not tomorrow?

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
```diff
@@ -430,6 +431,14 @@ void RemoteImageBufferProxy::prepareForBackingStoreChange()
 
+void RemoteImageBufferProxy::remoteBufferWasDestroyed()
+{
+    if (m_remoteBufferManager == RemoteBufferManager::Self && m_remoteRenderingBackendProxy)
```
I think the changes related to this are somehow out of place and hard to maintain.
I don't  immediately understand the  logic "if proxy is part of set, don't recreate, otherwise recreate".
I think the recreation logic should be understandable without understanding the sets.

To me the m_remoteBufferManager additions seem overly complicated and should be perhaps removed altogether. Definitively the name "Manager" to indicate "Yes, this is part of some set" is odd?

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBufferSetProxy.h:
```diff
@@ -0,0 +1,86 @@
+// Usage is done through RemoteRenderingBackendProxy::prepareBuffersForDisplay,
+// so that a Vector of RemoteRenderingBufferSets can be used with a single
+// IPC call.
+class RemoteRenderingBufferSetProxy : public RefCounted<RemoteRenderingBufferSetProxy>, public CanMakeWeakPtr<RemoteRenderingBufferSetProxy> {
```
I think this class should be RemoteImageBufferSetProxy.
Could this just reference the actual ImageBuffers (or RemoteImageBufferProxy instances) that it references?

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/PrepareBackingStoreBuffersData.serialization.in:
```diff
@@ -24,16 +24,26 @@
 
-[CustomHeader, AdditionalEncoder=StreamConnectionEncoder] struct WebKit::PrepareBackingStoreBuffersInputData {
-    WebKit::BufferIdentifierSet bufferSet;
+[CustomHeader] struct WebKit::PrepareBackingStoreBuffersInputData {
```
Now that you are importing a real existing concept, RemoteImageBufferSet, you should replace all references to the informal "BackingStoreBuffers" concept.

This could be called ImageBufferSetPrepareFrontBufferForDisplayInvocation with the exception that you come up with a better name.

See the toplevel comment for more.

kkinnunen-apple reviewed (commented):
So before, we had
RemoteRenderingBackend
 - Function1: assorted function making compositing work with ImageBuffers
 - Function2: assorted function making compositing work with ImageBuffers
 - ....
 

After, you import "image buffer set" concept to codify the above assorted things

 ImageBufferSet (Conceptual or actually exist as code)
   - Contains the operations that can be done to ImageBufferSet
   - updateConfiguration
   - prepareFrontBufferForDisplay
   
 RemoteImageBufferSetProxy
  - Reflection of what is ImageBufferSet, just a convenience that implements the similar api as the ImageBufferSet.
   - Mainly used for tracking the lifetime of the real ImageBuffer set in GPUP
   - I propose: holds the RemoteImageBufferProxy instances, like the ImageBufferSet holds ImageBuffer instances

 RemoteImageBufferSet
 - The IPC object that captures messages to the conceptual ImageBufferSet
 - Either forwards to concrete ImageBufferSet or just implements it if ImageBufferSet is not actual code artefact
 
 RemoteImageBufferSet messages
 - mainly the updateConfiguration message
 
RemoteRenderingBackend messages
- these are all messages of form:
  - here is the list of messages to apply to each ImageBufferSet
  - here is the list of replies that the ImageBufferSets gave.
 - So you could try to make the naming reflect this batch orientedness: there is no bespoke logic in this message, this is just a batch message to invoke batch member functions and return the result batch
 
FWIW, nothing that is actionable:
You moved the comment forward:  `// This is the GPU Process version of RemoteLayerBackingStore::prepareBuffers()`

Consider if we wanted to unify GPUP=off compositor and GPUP=on compositor.

Off: ImageBuffer
On: RemoteImageBuffer

Off: RemoteLayerBackingStore
On: RemoteImageBufferSet

So we might think that we should massage the classes of both On and Off to the point that GPUP is nothing more than RemoteXXProxy objects that send the messages, and RemoteXX objects that accept the message, and then XX objects that do the thing that is supposed to be done.. In this light, we are importing a new concept that doesn't exist in the Off case, and as such we are diverging from the goal of converging to the unified codebase. Of course, it's just something to think about, it of course cannot be such that the progress is to be made only with converging commits.

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
```diff
@@ -36,11 +36,15 @@ messages -> RemoteRenderingBackend NotRefCounted Stream {
+
 #if PLATFORM(COCOA)
-    PrepareBuffersForDisplay(Vector<WebKit::PrepareBackingStoreBuffersInputData> swapBuffersInput) -> (Vector<WebKit::PrepareBackingStoreBuffersOutputData> swapBuffersOutput) Synchronous NotStreamEncodableReply
+    CreateBuffersForDisplay(Vector<WebKit::PrepareBackingStoreBuffersInputData> swapBuffersInput)  -> (Vector<WebKit::PrepareBackingStoreBuffersOutputData> swapBuffersOutput) NotStreamEncodable NotStreamEncodableReply ReplyCanDispatchOutOfOrder
```
What was the reason for NotStreamEncodable?
I'd think that's only if we *send* MachPorts, and I don't think swapBuffersInput has those?

mattwoodrow replied:
Looks like I was just failing to put the right annotations in the serialisation generator file, fixed now.

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.cpp:
```diff
@@ -0,0 +1,258 @@
+    IntRect layerBounds = IntRect { { }, expandedIntSize(m_logicalSize) };
+
+    if (m_previousFrontBuffer && m_frontBuffer != m_previousFrontBuffer && !dirtyRegion.contains(layerBounds)) {
+        Region copyRegion(m_previouslyPaintedRect && 0  ? *m_previouslyPaintedRect : layerBounds);
```
some && 0 debug code?

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
```diff
@@ -187,6 +194,29 @@ void RemoteRenderingBackendProxy::releaseImageBuffer(RenderingResourceIdentifier
+
+void RemoteRenderingBackendProxy::releaseRemoteImageBufferSet(RemoteImageBufferSetProxy& bufferSet)
+{
+    auto iterator = m_bufferSets.find(bufferSet.remoteImageBufferSetIdentifier());
```
To be consistent, you could do remove(KeyType&) + ASSERT_UNUSED

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.h:
```diff
@@ -0,0 +1,94 @@
+    RemoteImageBufferSetProxy(RemoteRenderingBackendProxy&);
+    ~RemoteImageBufferSetProxy();
+
+    RemoteImageBufferSetIdentifier remoteImageBufferSetIdentifier() const { return m_identifier; }
```
I think this could be just `identifier()`

kkinnunen-apple commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
```diff
@@ -77,6 +77,9 @@ class RemoteLayerBackingStore : public CanMakeWeakPtr<RemoteLayerBackingStore> {
 #endif
 
+    virtual bool isRemoteLayerWithRemoteRenderingBackingStore() const { return false; }
+    virtual bool isRemoteLayerWithInProcessRenderingBackingStore() const { return false; }
```
Not your original problem but:  This split highlights the shortcoming of the ImageBuffer / RemoteRenderingBackend system.

There should be just one algorithm for RemoteLayerBackingStore, and the ImageBuffer abstraction and the needed factories and coordinator polymorphic types should provide the underlying implementation differences.

Maybe this commit might be considered as progression towards removing the distinction (unify two distinct  types with a base class, then remove the virtual functions one by one)

mattwoodrow replied:
That indeed would be nice, could add a base `ImageBufferSet` where `ImageBufferSetProxy` is a subclass.

kkinnunen-apple commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.h:
```diff
@@ -41,6 +41,8 @@ enum class SetNonVolatileResult : uint8_t;
 namespace WebKit {
 
 class RemoteLayerBackingStore;
+class RemoteLayerWithRemoteRenderingBackingStore;
```
FWIW: To me, with the words given, more understandable  would be:
```
RemoteLayerBackingStoreWithRemoteRendering
RemoteLayerBackingStoreWithInProcessRendering
```

Or even `RemoteLayerBackingStoreRemote`.
However, given how the concept uses `Remote` prefix to mean different thing than some other Remote uses somewhere, I see how that is a problematic naming.

This is one case where having just one `RemoteLayerBackingStore` would solve the naming issue. :)

kkinnunen-apple commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.h:
```diff
@@ -87,14 +90,21 @@ class RemoteLayerBackingStoreCollection : public CanMakeWeakPtr<RemoteLayerBacki
     virtual void markBackingStoreVolatileAfterReachabilityChange(RemoteLayerBackingStore&);
     virtual void markAllBackingStoreVolatileFromTimer();
 
+    bool collectRemoteRenderingBackingStoreBufferIdentifiersToMarkVolatile(RemoteLayerWithRemoteRenderingBackingStore&, OptionSet<VolatilityMarkingBehavior>, MonotonicTime now, Vector<WebCore::RenderingResourceIdentifier>&);
```
This is  a  mouthful. Maybe least the "identifier" in the middle would not be needed? E.g. if they were Pointers, you'd still not have "Pointer" in the name..

kkinnunen-apple commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.mm:
```diff
@@ -91,20 +97,17 @@
-        backingStore->applySwappedBuffers(nullptr, nullptr, nullptr, SwapBuffersDisplayRequirement::NeedsFullDisplay);
-        LOG_WITH_STREAM(RemoteLayerBuffers, stream << "RemoteLayerBackingStoreCollection::prepareBackingStoresForDisplay - allocated first buffers for layer " << backingStore->layer()->layerID() << " backing store " << *backingStore);
+        remoteRenderingBackend.prepareImageBufferSetsForDisplay(WTFMove(prepareBuffersData), [backingStoreList = WTFMove(backingStoreList)](Vector<RemoteRenderingBackendProxy::SwapBuffersResult>&& swapResult) {
+            RELEASE_ASSERT(swapResult.size() == backingStoreList.size());
```
When the async call is cancelled, this will release assert.
It'd be better to handle the error intentionally.

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.cpp:
```diff
@@ -0,0 +1,258 @@
+    auto bufferIdentifier = [](RefPtr<WebCore::ImageBuffer> buffer) -> std::optional<ProcessQualified<RenderingResourceIdentifier>> {
+        if (!buffer)
+            return std::nullopt;
+        return ProcessQualified(buffer->renderingResourceIdentifier(), Process::identifier());
```
I think this is that you replace the "generation" with process identifier?
I'm not 100% if we already transfer GPUP identifier to WP. However, I don't think it makes sense?
If we transfer 10000 image buffers, we keep on transferring the very same 64-bit id on an on, which could be somehow done differently without transferring this?
BTW This refactoring could have been made separately, so it'd be easier to review and land...

mattwoodrow replied:
I'll get rid of this for now, I don't think it adds much.

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
```diff
@@ -430,6 +320,12 @@ void RemoteImageBufferProxy::prepareForBackingStoreChange()
+void RemoteImageBufferProxy::remoteBufferWasDestroyed()
+{
+    if (m_remoteRenderingBackendProxy)
+        m_remoteRenderingBackendProxy->createRemoteImageBuffer(*this);
```
Is this change still related to this particular fix or some other fix?

mattwoodrow replied:
Leftover from previous changes, removed.

kkinnunen-apple commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.mm:
```diff
@@ -26,49 +26,100 @@
+        return;
+
+    if (!bufferSet->hasContext()) {
+        ASSERT(m_layer->owner()->platformCALayerDelegatesDisplay(m_layer));
```
is this missing early return?

mattwoodrow committed:
Sync wait in RemoteLayerWithRemoteRenderingBackingStoreCollection::prepareBackingStoresForDisplay delays rendering.
https://bugs.webkit.org/show_bug.cgi?id=265295
<rdar://116792212>

Reviewed by Kimmo Kinnunen.

In order to start issuing rendering commands for a layer, we need to determine which ImageBuffer is not
in-use and can be drawn to. We also need to know if the existing pixel contents have been discarded due
to being marked volatile. Both of these checks can only be done in the GPU process, so the previous
behaviour was to have a synchronous call to prepareBackingStoresForDisplay at the start of rendering.

This changes makes that process asynchronous (in the common case), by moving management of the buffer
triplet (front, back, secondary back) to the GPU process, and making the choice of which buffer is used
opaque from the perspective of the web process.

For each frame, the WebProcess allocates a new image buffer identifier (via RemoteDisplayListRecorderProxy),
and the GPU process asynchronously ensures that this is mapped to the right ImageBuffer (either by picking
an available one, or allocating if needed, similar to how recycling IOSurfaces from the pool happens). It
also queues copying of the pixels from the previous frame, clipping to the new draw region, and clearing the
draw region.

The only case where this isn't possible is when the existing front buffer has been marked volatile (and
might be discarded). This should be relatively rare (and not happen during high performance animations),
and we add a sync wait to determine if the full frame needs to be redrawn or not.

This adds a new RemoteRenderingBufferSet object in the GPU process that manages the set of buffers, and
preparing backing stores, volatility setting and flushing are changed to happen in terms of the buffer set
rather than individual ImageBuffers.

We unfortuntely still have to have a main-thread sync wait at the end of the rendering update, in order to
get the mach_port_t/IOSurface handle for the front buffer used. We also need to get the set of
RenderingResourceIdentifiers used, since these are the keys for the CAIOSurface caching in the UI process.
In the common case, this message will be sent back while the WebProcess is issuing drawing commands, and
the sync wait will process the pending message immediately.
Ideally a future change will handle sending IOSurfaces (via mach_port_t) from GPUP to UI process directly,
which would remove the need for this.

The current state is a performance improvement, since we can issue rendering commands before any sync
waits, but an ideal future would remove the sync waiting from rendering entirely.

This also fixes a bug with volatility, where we can send a volatility request, then do a sync prepare for
display (which sets non volatile), and finally process the asynchronous reply from the volatility request.
In that case the web process would cache locally that the buffer is volatile and not make further requests
for it. This is fixed by tracking transaction identifiers for volatility requests to ensure that we don't
get ordering problems.

* Source/WTF/wtf/text/TextStream.h:
(WTF::operator<<):
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.cpp:
(WebKit::RemoteImageBuffer::create):
(WebKit::RemoteImageBuffer::RemoteImageBuffer):
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.h:
(WebKit::RemoteImageBuffer::identifier const):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::didFailCreateImageBuffer):
(WebKit::RemoteRenderingBackend::didCreateImageBuffer):
(WebKit::RemoteRenderingBackend::moveToImageBuffer):
(WebKit::RemoteRenderingBackend::createImageBuffer):
(WebKit::RemoteRenderingBackend::createRemoteRenderingBufferSet):
(WebKit::RemoteRenderingBackend::releaseRemoteRenderingBufferSet):
(WebKit::RemoteRenderingBackend::updateRemoteRenderingBufferSetConfiguration):
(WebKit::RemoteRenderingBackend::createBuffersForDisplay):
(WebKit::RemoteRenderingBackend::markSurfacesVolatile):
(WebKit::handleFromBuffer): Deleted.
(WebKit::RemoteRenderingBackend::prepareBuffersForDisplay): Deleted.
(WebKit::RemoteRenderingBackend::prepareLayerBuffersForDisplay): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBufferSet.cpp: Added.
(WebKit::RemoteRenderingBufferSet::updateConfiguration):
(WebKit::RemoteRenderingBufferSet::ensureFrontBufferForDisplay):
(WebKit::RemoteRenderingBufferSet::prepareFrontBufferForDisplay):
(WebKit::RemoteRenderingBufferSet::makeBuffersVolatile):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBufferSet.h: Added.
(WebKit::RemoteRenderingBufferSet::isOpaque const):
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
(IPC::StreamClientConnection::waitForAsyncReplyAndDispatchImmediately):
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
(class_template_headers):
(headers_for_type):
* Source/WebKit/Shared/RemoteBufferSetIdentifier.h: Copied from Source/WebKit/WebProcess/GPU/graphics/BufferIdentifierSet.h.
* Source/WebKit/Shared/RemoteLayerTree/BufferAndBackendInfo.h: Removed.
* Source/WebKit/Shared/RemoteLayerTree/BufferAndBackendInfo.serialization.in: Removed.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
(WebKit::RemoteLayerBackingStore::dirtyRegion):
(WebKit::RemoteLayerBackingStore::setBufferCacheIdentifiers):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::ensureBackingStore):
(WebKit::RemoteLayerBackingStore::encode const):
(WebKit::RemoteLayerBackingStoreProperties::decode):
(WebKit::RemoteLayerBackingStoreProperties::dump const):
(WebKit::RemoteLayerBackingStore::setFrontBuffer):
(WebKit::RemoteLayerBackingStore::supportsPartialRepaint const):
(WebKit::RemoteLayerBackingStore::drawingRequiresClearedPixels const):
(WebKit::RemoteLayerBackingStore::setDelegatedContents):
(WebKit::RemoteLayerBackingStore::paintContents):
(WebKit::RemoteLayerBackingStore::drawInContext):
(WebKit::RemoteLayerBackingStoreProperties::applyBackingStoreToLayer):
(WebKit::RemoteLayerBackingStoreProperties::updateCachedBuffers):
(WebKit::RemoteLayerBackingStore::applySwappedBuffers): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.h:
(WebKit::RemoteLayerBackingStoreCollection::backingStoreUpdated):
(WebKit::RemoteLayerBackingStoreCollection::backingStoreWillBeEncoded):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStoreCollection.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStoreCollection.mm:
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::paintReachableBackingStoreContents):
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::prepareBackingStoresForDisplay):
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::allocateBufferForBackingStore):
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::collectBackingStoreBufferIdentifiersToMarkVolatile):
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::collectAllBufferIdentifiersToMarkVolatile):
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::markBackingStoreVolatileAfterReachabilityChange):
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::tryMarkAllBackingStoreVolatile):
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::markAllBackingStoreVolatileFromTimer):
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::sendMarkBuffersVolatile):
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::gpuProcessConnectionWasDestroyed):
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::backingStoreWasCreated):
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::backingStoreWillBeDestroyed):
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::backingStoreUpdated):
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::backingStoreWillBeEncoded):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.messages.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::asyncSetLayerContents):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::asyncSetLayerContents):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.h:
(WebKit::RemoteLayerTreeNode::asyncContentsIdentifier const):
(WebKit::RemoteLayerTreeNode::setAsyncContentsIdentifier):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/BufferIdentifierSet.h:
(WebKit::operator<<):
* Source/WebKit/WebProcess/GPU/graphics/BufferIdentifierSet.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/PrepareBackingStoreBuffersData.cpp:
(WebKit::operator<<):
* Source/WebKit/WebProcess/GPU/graphics/PrepareBackingStoreBuffersData.h:
* Source/WebKit/WebProcess/GPU/graphics/PrepareBackingStoreBuffersData.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::RemoteImageBufferProxy):
(WebKit::m_remoteBufferManager):
(WebKit::RemoteImageBufferProxy::remoteBufferWasDestroyed):
(WebKit::RemoteSerializedImageBufferProxy::sinkIntoImageBuffer):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
(WebKit::RemoteImageBufferProxy::create):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::didClose):
(WebKit::RemoteRenderingBackendProxy::createImageBuffer):
(WebKit::RemoteRenderingBackendProxy::createRemoteRenderingBufferSet):
(WebKit::RemoteRenderingBackendProxy::releaseRemoteRenderingBufferSet):
(WebKit::RemoteRenderingBackendProxy::updateRemoteRenderingBufferSetConfiguration):
(WebKit::RemoteRenderingBackendProxy::prepareBuffersForDisplay):
(WebKit::RemoteRenderingBackendProxy::ensurePrepareCompleted):
(WebKit::RemoteRenderingBackendProxy::markSurfacesVolatile):
(WebKit::RemoteRenderingBackendProxy::didMarkLayersAsVolatile):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBufferSetProxy.cpp: Copied from Source/WebKit/WebProcess/GPU/graphics/BufferIdentifierSet.h.
(WebKit::RemoteRenderingBufferSetProxy::RemoteRenderingBufferSetProxy):
(WebKit::RemoteRenderingBufferSetProxy::~RemoteRenderingBufferSetProxy):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBufferSetProxy.h: Added.
(WebKit::RemoteRenderingBufferSetProxy::remoteBufferSetIdentifier const):
(WebKit::RemoteRenderingBufferSetProxy::requestedVolatility):
(WebKit::RemoteRenderingBufferSetProxy::confirmedVolatility):
(WebKit::RemoteRenderingBufferSetProxy::addRequestedVolatility):
(WebKit::RemoteRenderingBufferSetProxy::setConfirmedVolatility):
(WebKit::RemoteRenderingBufferSetProxy::clearVolatilityUntilAfter):
(WebKit::operator<<):
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::remoteResourceCacheWasDestroyed):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDocumentLayout.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.mm:

Canonical link: https://commits.webkit.org/271747@main
